src/FormSpy.js

```diff
@@ -1,194 +1,44 @@
-        this.unsubscribe()
-        this.subscribe(this.props, this.notify)
-      }
-    }
-  }
-
-  componentWillUnmount() {
-    this.unsubscribe()
-  }
-
-  render() {
-    const { onChange, reactFinalForm, ...rest } = this.props
-    const renderProps: FormSpyRenderProps = {
-      batch:
-        reactFinalForm &&
-        ((fn: () => void) => {
-          // istanbul ignore next
-          if (process.env.NODE_ENV !== 'production') {
-            console.error(
-              `Warning: As of React Final Form v3.3.0, props.batch() is deprecated and will be removed in the next major version of React Final Form. Use: props.form.batch() instead. Check your FormSpy render prop.`
-            )
-          }
-          return reactFinalForm.batch(fn)
-        }),
-      blur:
-        reactFinalForm &&
-        ((name: string) => {
-          // istanbul ignore next
-          if (process.env.NODE_ENV !== 'production') {
-            console.error(
-              `Warning: As of React Final Form v3.3.0, props.blur() is deprecated and will be removed in the next major version of React Final Form. Use: props.form.blur() instead. Check your FormSpy render prop.`
-            )
-          }
-          return reactFinalForm.blur(name)
-        }),
-      change:
-        reactFinalForm &&
-        ((name: string, value: any) => {
-          // istanbul ignore next
-          if (process.env.NODE_ENV !== 'production') {
-            console.error(
-              `Warning: As of React Final Form v3.3.0, props.change() is deprecated and will be removed in the next major version of React Final Form. Use: props.form.change() instead. Check your FormSpy render prop.`
-            )
-          }
-          return reactFinalForm.change(name, value)
-        }),
-      focus:
-        reactFinalForm &&
-        ((name: string) => {
-          // istanbul ignore next
-          if (process.env.NODE_ENV !== 'production') {
-            console.error(
-              `Warning: As of React Final Form v3.3.0, props.focus() is deprecated and will be removed in the next major version of React Final Form. Use: props.form.focus() instead. Check your FormSpy render prop.`
-            )
-          }
-          return reactFinalForm.focus(name)
-        }),
-      form: {
-        ...reactFinalForm,
-        reset: eventOrValues => {
-          if (isSyntheticEvent(eventOrValues)) {
-            // it's a React SyntheticEvent, call reset with no arguments
-            reactFinalForm.reset()
-          } else {
-            reactFinalForm.reset(eventOrValues)
-          }
+  const state = useFormState({ onChange, subscription })
+  const renderProps: FormSpyRenderProps = {
+    form: {
+      ...reactFinalForm,
+      reset: eventOrValues => {
+        if (isSyntheticEvent(eventOrValues)) {
+          // it's a React SyntheticEvent, call reset with no arguments
+          reactFinalForm && reactFinalForm.reset()
```

Comment:
`reactFinalForm` cant be undefined here, I think Flow should understand that - but might be wrong

Reply:
Good catch! It [_was_ undefined-able (?) once](https://twitter.com/erikras/status/1127989923334754315), but I engineered a way around it.

Reply:
Well, it _can_ be if the developer is a doofus and tries to put a `<FormSpy>` outside of a `<Form>`, hence the warning above. Flow is in the right here.

Reply:
Oh, I thought you throw error above - wouldnt it be better to throw as it's an obvious mistake and you wouldnt have to guard against missing value all over the place?

Reply:
Good point. I decided to make `ReactFinalForm` throw errors; so I'm going to make them all throw if you're using it incorrectly, production or not.